### PR TITLE
CGAL_CORE:  Make it thread safe

### PR DESCRIPTION
**Pull request by afabri**
Towards a fix of Issue #1397

### CODE REVIEW

**Comment by lrineau** [comment]
There are runtime errors in `test_LEDA_IO.cpp`  in the Installation line:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-86.shtml#Installation

And also several new errors in Algebraic_kernel_d:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-86.shtml#Algebraic_kernel_d

This is probably due to this PR.

**Comment by sloriot** [comment]
ping @afabri

**Comment by sloriot** [comment]
Errors in `Installation` are in [master](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-I-90.shtml#Installation) and the one in `Algebraic_kernel_d` too.

**Comment by lrineau** [comment]
The problem in `master` for `test_LEDA_io` are reported in https://github.com/CGAL/cgal/issues/1788.

As for Algebraic_kernel, if I compare `master` (I-96) with the last `integration` with this PR (Ic-86):

```
--- CGAL-4.10-I-96/results_afabri_x64_Cygwin-Windows8_MSVC2012-Release-64bits.txt       2017-01-09 02:02:05.000000000 +0100
+++ CGAL-4.10-Ic-86/results_afabri_x64_Cygwin-Windows8_MSVC2012-Release-64bits.txt      2016-12-28 01:52:56.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
--- CGAL-4.10-I-96/results_afabri_x64_Cygwin-Windows8_MSVC2013-Release-32bits.txt       2017-01-09 09:19:22.000000000 +0100
+++ CGAL-4.10-Ic-86/results_afabri_x64_Cygwin-Windows8_MSVC2013-Release-32bits.txt      2016-12-28 08:21:21.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
--- CGAL-4.10-I-96/results_cgaltester_x86-64_Darwin-13.0_Apple-clang-5.0_Release-cpp11.txt      2017-01-09 03:59:20.000000000 +0100
+++ CGAL-4.10-Ic-86/results_cgaltester_x86-64_Darwin-13.0_Apple-clang-5.0_Release-cpp11.txt     2016-12-28 04:05:06.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
--- CGAL-4.10-I-96/results_cgaltester_x86-64_Darwin-13.0_Apple-clang-5.0_Release.txt    2017-01-09 00:47:30.000000000 +0100
+++ CGAL-4.10-Ic-86/results_cgaltester_x86-64_Darwin-13.0_Apple-clang-5.0_Release.txt   2016-12-28 00:49:47.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
--- CGAL-4.10-I-96/results_lrineau_ArchLinux-clang.txt  2017-01-09 06:19:43.000000000 +0100
+++ CGAL-4.10-Ic-86/results_lrineau_ArchLinux-clang.txt 2016-12-28 10:13:59.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
--- CGAL-4.10-I-96/results_lrineau_CentOS6-32.txt       2017-01-09 09:43:04.000000000 +0100
+++ CGAL-4.10-Ic-86/results_lrineau_CentOS6-32.txt      2016-12-28 12:59:31.000000000 +0100
-Algebraic_kernel_d r
+Algebraic_kernel_d n
```

Six cells turned to red. That is really significant.

**Comment by lrineau** [comment]
@afabri Note that I have rebased this PR onto `master`, to get all merged thread-safety modifications.

**Comment by lrineau** [comment]
There are still runtime errors in Polynomial, for Apple-clang, for example:
```
------------------------------------------------------------------
- ProgramOutput.Interpolator.x86-64_Darwin-13.0_Apple-clang-5.0_Release
------------------------------------------------------------------
CGAL::Random()::get_seed() = 1484001755
N4CORE9BigIntRepE
CGAL error: assertion violation!
Expression : ! blocks.empty()
File       : /Users/cgaltester/cgal_test/CGAL-4.10-Ic-97/include/CGAL/CORE/MemoryPool.h
Line       : 127
Explanation: 
Refer to the bug-reporting instructions at http://www.cgal.org/bug_report.html
libc++abi.dylib: terminating with unexpected exception of type CGAL::Assertion_exception: CGAL ERROR: assertion violation!
Expr: ! blocks.empty()
File: /Users/cgaltester/cgal_test/CGAL-4.10-Ic-97/include/CGAL/CORE/MemoryPool.h
Line: 127
```
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-97.shtml#Polynomial

File: Installation/include/CGAL/tss.h [hunk 1]
```diff
@@ -59,6 +67,8 @@
 
 #  define CGAL_STATIC_THREAD_LOCAL_VARIABLE(TYPE, VAR,ARG1) static TYPE VAR(ARG1)
 
+#  define CGAL_STATIC_THREAD_LOCAL_VARIABLE(TYPE, VAR,ARG1,ARG2) static TYPE VAR(ARG1,ARG2)
```
Comment (lrineau) [inline Installation/include/CGAL/tss.h]: _2

**Pull request merged**